File: week1/move-zeros.java
class Solution {
    public void moveZeroes(int[] nums) {
        if (nums == null) {
            return;
        }
        // Solution 2, double pointers i and j, moving non-zero to left, zero to right, like quick sort
        int j = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                int tmp = nums[i];
                nums[i] = nums[j];
                nums[j++] = tmp;
            }
        }

        // // Solution 1, moving non-zero nums forward, then appending zero to tail
        // int j = 0;
        // for (int i = 0; i < nums.length; i++) {
        //     if (nums[i] != 0) {
        //         // move non-zero forward
        //         nums[j++] = nums[i];
        //     }
        // }
        
        // // append zeros
        // for (int i = j; i < nums.length; i++) {
        //     nums[i] = 0; 
        // }
    }
}

// Reference
// https://leetcode-cn.com/problems/move-zeroes/solution/dong-hua-yan-shi-283yi-dong-ling-by-wang_ni_ma/

File: week1/move-zeros.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int size = nums.size();
        if (size == 0) {
            return;
        }

        // Solution 2
        int j = 0;
        for (int i = 0; i < size; i++) {
            if (nums[i] != 0) {
                swap(nums[i], nums[j]);
                j++;
            }
        }

        // Solution 1
        // int cur = 0;
        // for (int i = 0; i < size; i++) {
        //     if (nums[i] != 0) {
        //         nums[cur] = nums[i];
        //         ++cur;
        //     }
        // }

        // while (cur < size) {
        //     nums[cur] = 0;
        //     ++cur;
        // }
    }
};